### PR TITLE
Add Password for login method in github_release

### PR DESCRIPTION
**Pull request by Akasurde**
##### SUMMARY
Fix allows user to specify username and password for
using github_release to perform various operations

Signed-off-by: Abhijeet Kasurde <akasurde@redhat.com>

##### ISSUE TYPE
 - Bugfix Pull Request

##### COMPONENT NAME
lib/ansible/modules/source_control/github_release.py

##### ANSIBLE VERSION
```
2.4 devel
```

### CODE REVIEW

File: lib/ansible/modules/source_control/github_release.py [hunk 1]
```diff
@@ -96,19 +114,26 @@ def main():
...
     repo = module.params['repo']
     user = module.params['user']
+    password = module.params['password']
     login_token = module.params['token']
     action = module.params['action']
 
     # login to github
     try:
-        gh = github3.login(token=str(login_token))
+        if user and password:
```
Comment (adrianmoisey) [inline lib/ansible/modules/source_control/github_release.py:123]: Just a thought, and this was a problem I created, `user` was meant to be used of the repository that needs to be accessed, not for auth.
I think we should change `repo` to be the full path of the repository (`username/repo`) and then we can split it for use later on.

File: lib/ansible/modules/source_control/github_release.py [hunk 2]
```diff
@@ -77,13 +93,15 @@
     HAS_GITHUB_API = True
 except ImportError:
     HAS_GITHUB_API = False
+from ansible.module_utils.basic import AnsibleModule, get_exception
```
Comment (adrianmoisey) [inline lib/ansible/modules/source_control/github_release.py:94]: ❤️

**Comment by adrianmoisey** [comment]
shipit

File: lib/ansible/modules/source_control/github_release.py (hunk 1 above)
Comment (adrianmoisey) [reply lib/ansible/modules/source_control/github_release.py:123]: This is something that can be fixed/changed in another PR

Comment (Akasurde) [reply lib/ansible/modules/source_control/github_release.py:123]: @adrianmoisey Yes. Sure I will raise another PR for that.

File: lib/ansible/modules/source_control/github_release.py (hunk 2 above)
Comment (Akasurde) [reply lib/ansible/modules/source_control/github_release.py:94]: @adrianmoisey Thanks.

**Comment by Akasurde** [comment]
ready_for_review

**Comment by Akasurde** [comment]
ready_for_review

File: lib/ansible/modules/source_control/github_release.py [hunk 3]
```diff
@@ -34,14 +34,23 @@
...
+        default: null
     user:
         required: true
         description:
             - The GitHub account that owns the repository
+        default: null
+    password:
+        required: true
+        description:
+            - The Github account password for the user
```
Comment (gundalow) [inline lib/ansible/modules/source_control/github_release.py]: GitHub

This applies in a number of places.

File: lib/ansible/modules/source_control/github_release.py [hunk 4]
```diff
@@ -77,13 +93,15 @@
...
     HAS_GITHUB_API = False
+from ansible.module_utils.basic import AnsibleModule, get_exception
 
 
 def main():
     module = AnsibleModule(
         argument_spec=dict(
             repo=dict(required=True),
             user=dict(required=True),
+            password=dict(required=True),
```
Comment (gundalow) [inline lib/ansible/modules/source_control/github_release.py]: Please add `, no_log=True`

Comment (gundalow) [reply lib/ansible/modules/source_control/github_release.py]: Is this actually a required option?

1) Are `password` and `token` mutually exclusive. 
2) can you use this without auth? 


There are some options to `AnsibleModule` which you may be able to use to validate options passed in.
Look at existing modules for examples:
```
mutually_exclusive
required_one_of
```

File: lib/ansible/modules/source_control/github_release.py [hunk 5]
```diff
@@ -34,14 +34,23 @@
...
         description:
             - Github Personal Access Token for authenticating
+        default: null
     user:
         required: true
         description:
             - The GitHub account that owns the repository
+        default: null
+    password:
+        required: true
```
Comment (gundalow) [inline lib/ansible/modules/source_control/github_release.py]: Based on your examples I don't believe this is a required option, so this lime can be removed.

File: lib/ansible/modules/source_control/github_release.py [hunk 6]
```diff
@@ -56,11 +65,18 @@
...
 EXAMPLES = '''
 - name: Get latest release of test/test
-  github:
+  github_release:
     token: tokenabc1234567890
     user: testuser
     repo: testrepo
     action: latest_release
+
+- name: Get latest release of test repo using username and password
```
Comment (gundalow) [inline lib/ansible/modules/source_control/github_release.py]: `Get latest release of test repo using username and password. Ansible 2.4.`

**Review by gundalow — CHANGES_REQUESTED** [review]

**Comment by gundalow** [comment]
Closing and reopening to rerun CI.

File: lib/ansible/modules/source_control/github_release.py (hunk 3 above)
Comment (Akasurde) [reply lib/ansible/modules/source_control/github_release.py]: Done

File: lib/ansible/modules/source_control/github_release.py (hunk 6 above)
Comment (Akasurde) [reply lib/ansible/modules/source_control/github_release.py]: Done.

File: lib/ansible/modules/source_control/github_release.py (hunk 5 above)
Comment (Akasurde) [reply lib/ansible/modules/source_control/github_release.py]: Done.

File: lib/ansible/modules/source_control/github_release.py (hunk 4 above)
Comment (Akasurde) [reply lib/ansible/modules/source_control/github_release.py]: Done.

File: lib/ansible/modules/source_control/github_release.py [hunk 7]
```diff
@@ -77,17 +93,21 @@
...
         argument_spec=dict(
             repo=dict(required=True),
             user=dict(required=True),
+            password=dict(required=True, no_log=True),
             token=dict(required=True, no_log=True),
             action=dict(required=True, choices=['latest_release']),
         ),
-        supports_check_mode=True
+        supports_check_mode=True,
+        required_one_of=(('password', 'token'),),
```
Comment (gundalow) [inline lib/ansible/modules/source_control/github_release.py:107]: I believe you need to remove `required=True` from `password` and `token` in `argument_spec` and in the `DOCUMENTATION` block.

**Review by gundalow — APPROVED** [review]

**Comment by gundalow** [comment]
@adrianmoisey There have been a few changes since you looked last, could you please cast your eye over this again. Thanks

**Comment by Akasurde** [comment]
@adrianmoisey Could you please take a look into this PR ?

**Comment by Akasurde** [comment]
@adrianmoisey Ping.

**Comment by adrianmoisey** [comment]
Sorry, I've been missing the notification.
This looks great! :+1:

**Comment by adrianmoisey** [comment]
shipit

**Comment by adrianmoisey** [comment]
(I don't know how this bot works, or if I'm considered a maintainer)

**Comment by gundalow** [comment]
Closing/re-opening to trigger CI

**Comment by gundalow** [comment]
Merged, thanks :)

**Comment by Akasurde** [comment]
@gundalow Thanks for review.

**Comment by Akasurde** [comment]
@adrianmoisey Thanks for review.

**Pull request merged**